src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -4166,16 +4167,208 @@ private void ImportCompareOperation(ILOpcode opcode)
+                    AddInternalBasicBlock(noUnderflowBlock);
+                }
+            }
+            else
+            {
+                LLVMValueRef cmp;
+                //special case same width signed -> unsigned, can just check for negative values
+                if (unsigned && (loadedValue.TypeOf.IntWidth >> 3) == destType.InstanceFieldSize.AsInt && 
+                    (sourceType == GetWellKnownType(WellKnownType.Int16)
+                    || sourceType == GetWellKnownType(WellKnownType.Int32)
+                    || sourceType == GetWellKnownType(WellKnownType.Int64)))
+                {
+                    cmp = _builder.BuildICmp(LLVMIntPredicate.LLVMIntSLT, loadedValue, LLVMValueRef.CreateConstInt(loadedValue.TypeOf, 0));
+                }
+                else
+                {
+                    var overflowBits = _builder.BuildNot(mask);
+                    var check = _builder.BuildAnd(overflowBits, loadedValue);
+                    cmp = _builder.BuildICmp(LLVMIntPredicate.LLVMIntNE, LLVMValueRef.CreateConstInt(loadedValue.TypeOf, 0), check);
+                }
+
+                LLVMBasicBlockRef overflowBlock = _currentFunclet.AppendBasicBlock("ovf");
+                LLVMBasicBlockRef noOverflowBlock = _currentFunclet.AppendBasicBlock("no_ovf");
+                _builder.BuildCondBr(cmp, overflowBlock, noOverflowBlock);
+
+                _builder.PositionAtEnd(overflowBlock);
+                CallOrInvokeThrowException(_builder, "ThrowHelpers", "ThrowOverflowException");
+                _builder.PositionAtEnd(noOverflowBlock);
+                AddInternalBasicBlock(noOverflowBlock);
+            }
+        }
+
+        private void BuildNegativeConvChecks(LLVMValueRef loadedValue, long minValue, EcmaType destType)
+        {
+            var mask = LLVMValueRef.CreateConstInt(loadedValue.TypeOf, (ulong)minValue);
+            if (IsLlvmReal(loadedValue.TypeOf))
+            {
+                var cmp = _builder.BuildFCmp(LLVMRealPredicate.LLVMRealULE, loadedValue, MaximumNegativeFailingDouble(loadedValue.TypeOf, minValue - 1, destType));  // <= min - 1 as rounds to 0, e.g. -128.9 is allowed in sbyte
+                LLVMBasicBlockRef overflowBlock = _currentFunclet.AppendBasicBlock("unff");
+                LLVMBasicBlockRef noOverflowBlock = _currentFunclet.AppendBasicBlock("no_unff");
+                _builder.BuildCondBr(cmp, overflowBlock, noOverflowBlock);
+            
+                _builder.PositionAtEnd(overflowBlock);
+                CallOrInvokeThrowException(_builder, "ThrowHelpers", "ThrowOverflowException");
+                _builder.PositionAtEnd(noOverflowBlock);
+                AddInternalBasicBlock(noOverflowBlock);
+            }
+            else
+            {
+                var overflowBits = _builder.BuildNot(mask);
+                var check = _builder.BuildXor(overflowBits, loadedValue); // for no overflow the mask will clear all values and the result should be all 0s
```

Comment:
I am having hard time following what this masking does.

I would expect this to be a simple:

`if ((unsigned)(x - minValue) >= (unsigned)(maxValue - minValue)) ThrowOverflowException();`

Are there good reasons for doing anything else?

Reply:
Not really, the masking of the most significant bits that are getting lost was just what came to mind.  I'll change it for clarity.